snakemake/executors/flux.py

```diff
@@ -0,0 +1,242 @@
+        fields2attrs = {
+            "id.f58": (),
+            "username": ("userid",),
+            "exception.occurred": ("exception_occurred",),
+            "exception.type": ("exception_type",),
+            "exception.note": ("exception_note",),
+            "runtime": ("t_run", "t_cleanup"),
+            "status": ("state", "result"),
+            "status_abbrev": ("state", "result"),
+            "t_remaining": ("expiration", "state", "result"),
+        }
+
+        # Set job attributes we will use later to get job statuses
+        self.job_attrs = set()
+        for field in self.jobs_formatter.fields:
+            if field not in fields2attrs:
+                self.job_attrs.update((field,))
+            else:
+                self.job_attrs.update(fields2attrs[field])
+
+    def _init_flux(self):
+        """
+        Ensure we can import flux in the environment.
+        """
+        if not flux:
+            sys.exit(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
+            )
+        self._fexecutor = flux.job.FluxExecutor()
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def format_job_exec(self, job):
+        """
+        We don't want a python -m to run.
+        """
+        prefix = self.get_job_exec_prefix(job)
+        if prefix:
+            prefix += " &&"
+        suffix = self.get_job_exec_suffix(job)
+        if suffix:
+            suffix = f"&& {suffix}"
+        return join_cli_args(
+            [
+                prefix,
+                self.get_envvar_declarations(),
+                "snakemake",
+                format_cli_arg("--snakefile", self.get_snakefile()),
+                self.get_job_args(job),
+                self.general_args,
+                suffix,
+            ]
+        )
+
+    def run(self, job, callback=None, submit_callback=None, error_callback=None):
+        """
+        Submit a job to flux.
+        """
+        super()._run(job)
+
+        from flux.job import JobspecV1
```

Comment:
Is this import necessary, given that `__init__` throws an error if `flux.job` can't be imported? Is `run()` ever invoked as a free function rather than an instance method?

Reply:
I remember there being something buggy going on where if I didn't have this line, it wouldn't work. So for now I think it's safer to be conservative and leave, if that's OK!

Reply:
This is updated to a WorkflowError - I think it could also be an ImportError, but it does reflect an issue with the workflow running.

Reply:
Not sure I understand your last comment. But regarding your first comment, if you _just_ removed the `from flux.job import JobspecV1` line I would expect it to fail, you would also need to change the `JobspecV1(...)` calls to `flux.job.JobspecV1(...)` calls. In any case not a big deal though.

Reply:
Yep wrong thread (sorry!) I don't remember the issue, but I know I put it there like that for a specific reason.